libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/TwoPhaseValidation.hs

```diff
@@ -1610,303 +1607,686 @@ type A = AlonzoEra C_Crypto
+  evaluateTransactionFee @A
+    pparams
+    validatingTxNoWits
+    1
+    @?= minfee pparams (validatingTx pf)
+  where
+    pf = Alonzo Mock
+    pparams = newPParams pf $ defaultPPs ++ [MinfeeA 1]
+    validatingTxNoWits =
+      newTx
+        pf
+        [ Body (validatingBody pf),
+          WitnessesI
+            [ ScriptWits' [always 3 pf],
+              DataWits' [datumExample1],
+              RdmrWits validatingRedeemersEx1
+            ]
+        ]
+
+alonzoAPITests :: TestTree
+alonzoAPITests =
+  testGroup "Alonzo API" [testCase "evaluateTransactionFee" testEvaluateTransactionFee]
+
+-- =====================================================================================
+-- Proof parameterized TestTrees
+
+hashScriptIntegrityByProof ::
+  Proof era ->
+  Core.PParams era ->
+  Set.Set Language ->
+  Redeemers era ->
+  TxDats era ->
+  StrictMaybe (ScriptIntegrityHash (Crypto era))
+hashScriptIntegrityByProof (Alonzo _) = hashScriptIntegrity
+hashScriptIntegrityByProof (Babbage _) = hashScriptIntegrity
+hashScriptIntegrityByProof _ = \_ _ _ _ -> SNothing
+
+-- | This type is what you get when you use runSTS in the UTXOW rule. It is also
+--   the type one uses for expected answers, to compare the 'computed' against 'expected'
+type Result era = Either [(PredicateFailure (Core.EraRule "UTXOW" era))] (State (Core.EraRule "UTXOW" era))
+
+testUTXOWwith ::
+  forall era.
+  ( GoodCrypto (Crypto era),
+    Default (State (EraRule "PPUP" era)),
+    PostShelley era
+  ) =>
+  WitRule "UTXOW" era ->
+  (Result era -> Result era -> Assertion) ->
+  Core.PParams era ->
+  Core.Tx era ->
+  Result era ->
+  Assertion
+testUTXOWwith wit@(UTXOW proof) cont pparams tx expected =
+  let env = utxoEnv pparams
+      state = initialUtxoSt proof
+   in case proof of
+        Alonzo _ -> runSTS wit (TRC (env, state, tx)) (cont expected)
+        Babbage _ -> runSTS wit (TRC (env, state, tx)) (cont expected)
+        Mary _ -> runSTS wit (TRC (env, state, tx)) (cont expected)
+        Allegra _ -> runSTS wit (TRC (env, state, tx)) (cont expected)
+        Shelley _ -> runSTS wit (TRC (env, state, tx)) (cont expected)
+
+-- | A small example of what a continuation for 'runSTS' might look like
+genericCont ::
+  ( Eq x,
+    Show x,
+    PrettyA x,
+    Eq y,
+    Show y,
+    PrettyA y
+  ) =>
+  Either [x] y ->
+  Either [x] y ->
+  Assertion
+genericCont expected computed =
+  case (computed, expected) of
+    (Left c, Left e) -> c @?= e
+    (Right c, Right e) -> c @?= e
+    (Left x, Right y) ->
+      error $
+        "expected to pass with "
+          ++ show (prettyA y)
+          ++ "\n\nBut failed with\n\n"
+          ++ show (ppList prettyA x)
+    (Right x, Left y) ->
+      error $
+        "expected to fail with "
+          ++ show (ppList prettyA y)
+          ++ "\n\nBut passed with\n\n"
+          ++ show (prettyA x)
+
+isSubset :: Eq t => [t] -> [t] -> Bool
+isSubset small big = List.all (`List.elem` big) small
```

Comment:
This is quite expensive. Can we rely on `Ord` instead of  just `Eq`
```suggestion
isSubset :: Eq t => [t] -> [t] -> Bool
isSubset small big = List.all (`Set.member` bigSet) small
  where bigSet = Set.fromList big
```

Reply:
Typically these things have a length of 1-4, so I am not too worried. This is in the tests, and the current tests run in less than a second. I don't think that will change.